fdbserver/EncryptKeyProxy.actor.cpp

```diff
@@ -18,53 +18,312 @@
+	    baseCipherKeyIdCacheMisses("EKPCipherIdCacheMisses", ekpCacheMetrics),
+	    baseCipherKeyIdCacheHits("EKPCipherIdCacheHits", ekpCacheMetrics),
+	    baseCipherDomainIdCacheMisses("EKPCipherDomainIdCacheMisses", ekpCacheMetrics),
+	    baseCipherDomainIdCacheHits("EKPCipherDomainIdCacheHits", ekpCacheMetrics),
+	    baseCipherKeysRefreshed("EKPCipherKeysRefreshed", ekpCacheMetrics) {}
 
-	explicit EncryptKeyProxyData(UID id) : myId(id) {}
+	void insertIntoBaseDomainIdCache(const EncryptDomainId domainId,
+	                                 const EncryptBaseCipherId baseCipherId,
+	                                 const StringRef baseCipherKey) {
+		// Entries in domainId cache are eligible for periodic refreshes to support 'limiting lifetime of encryption
+		// key' support if enabled on external KMS solutions.
+
+		baseCipherDomainIdCache[domainId] = EncryptBaseCipherKey(domainId, baseCipherId, baseCipherKey, false);
+
+		// Update cached the information indexed using baseCipherId
+		insertIntoBaseCipherIdCache(domainId, baseCipherId, baseCipherKey);
+	}
+
+	void insertIntoBaseCipherIdCache(const EncryptDomainId domainId,
+	                                 const EncryptBaseCipherId baseCipherId,
+	                                 const StringRef baseCipherKey) {
+		// Given an cipherKey is immutable, it is OK to NOT expire cached information.
+		// TODO: Update cache to support LRU eviction policy to limit the total cache size.
+
+		baseCipherKeyIdCache[baseCipherId] = EncryptBaseCipherKey(domainId, baseCipherId, baseCipherKey, true);
+	}
 };
 
+ACTOR Future<Void> getCipherKeysByBaseCipherKeyIds(Reference<EncryptKeyProxyData> ekpProxyData,
+                                                   SimKmsProxyInterface simKmsInterface,
+                                                   EKPGetBaseCipherKeysByIdsRequest req) {
+	// Scan the cached cipher-keys and filter our baseCipherIds locally cached
+	// for the rest, reachout to KMS to fetch the required details
+
+	std::vector<EncryptBaseCipherId> lookupCipherIds;
+	state std::unordered_map<EncryptBaseCipherId, Standalone<StringRef>> cachedKeys;
+
+	for (EncryptBaseCipherId id : req.baseCipherIds) {
+		const auto itr = ekpProxyData->baseCipherKeyIdCache.find(id);
+		if (itr != ekpProxyData->baseCipherKeyIdCache.end()) {
+			ASSERT(itr->second.isValid());
+			cachedKeys.emplace(id, itr->second.baseCipherKey);
+		} else {
+			lookupCipherIds.push_back(id);
+		}
+	}
+
+	ekpProxyData->baseCipherKeyIdCacheHits += cachedKeys.size();
+	ekpProxyData->baseCipherKeyIdCacheMisses += lookupCipherIds.size();
+
+	state EKPGetBaseCipherKeysByIdsRequest keysByIds = req;
+	state EKPGetBaseCipherKeysByIdsReply keyIdsReply;
+
+	if (g_network->isSimulated()) {
+		if (!lookupCipherIds.empty()) {
+			try {
+				SimGetEncryptKeysByKeyIdsRequest simKeyIdsReq(lookupCipherIds);
+				SimGetEncryptKeysByKeyIdsReply simKeyIdsReply =
+				    wait(simKmsInterface.encryptKeyLookupByKeyIds.getReply(simKeyIdsReq));
```

Comment:
Since the kms calls are simulated and in the same process this should be fine, but against an actual KMS you'll want to have error + timeout handling

Reply:
Good point. 
I'm still debating whether to let the caller of EKPServer handles the retires OR EKPServer should retry on the callers behalf.